upgrading/topics/keycloak/changes-19_0_0.adoc

```diff
@@ -0,0 +1,216 @@
+= Changes affecting developers
+
+{project_name} undergoes large refactoring which impacts existing code.
+Some of these changes require updates to existing code.
+These are in more detailed described below.
+
+== Rationale for changes
+
+{project_name} has several limitations; for example, downtime is needed for upgrading a {project_name} cluster.
+To address the limitations, in-depth refactor has been initiated.
+
+The changes in this version are mostly attached to storage refactoring and a preparation of a new storage, called map storage, which will eventually replace the current storage, which with this version becomes called a _legacy store_.
+The legacy store will still be available in {project_name} for several more versions.
+
+The new store imposes a strict separation of responsibility between the service and storage layers.
+For that reason, the service layer's visibility of an object's origin will be restricted, so it will not be able to discriminate between cached or non-cached objects, or objects originating from local or federated storage.
+
+User storage SPI will become deprecated.
+It will be supported for several more versions, but will be eventually replaced by map storage SPI which will offer the ability to create custom storages for any recognized area, e.g. users, roles, clients or groups.
```

Comment:
```suggestion
It will be supported for several more versions, but will be eventually replaced by the Map Storage SPI, which will offer the ability to create custom storages for any recognized area, such as users, roles, clients or groups.
```